evap/staff/importers/enrollment.py

```diff
@@ -0,0 +1,652 @@
+
+        user_data_list, course_data_list = normalize_rows(parsed_rows, importer_log)
+        existing_user_profiles, new_user_profiles = get_user_profile_objects(user_data_list)
+
+        responsible_emails = {course_data.responsible_email for course_data in course_data_list}
+        new_responsibles_count = sum(1 for obj in new_user_profiles if obj.email in responsible_emails)
+        new_participants_count = len(new_user_profiles) - new_responsibles_count
+
+        importer_log.raise_if_has_errors()
+        if test_run:
+            importer_log.add_success(_("The test run showed no errors. No data was imported yet."))
+            msg = format_html(
+                _("The import run will create {evaluation_count} courses/evaluations and {user_count} users:"),
+                evaluation_count=len(course_data_list),
+                user_count=len(new_user_profiles),
+            )
+            msg += create_user_list_html_string_for_message(new_user_profiles)
+            importer_log.add_success(msg)
+        else:
+            update_existing_and_create_new_user_profiles(existing_user_profiles, new_user_profiles)
+            store_course_data_in_db(course_data_list, semester, vote_start_datetime, vote_end_date)
+            store_participants_in_db(parsed_rows)
+
+            msg = format_html(
+                _("Successfully created {} courses/evaluations, {} participants and {} contributors:"),
+                len(course_data_list),
+                new_participants_count,
+                new_responsibles_count,
+            )
+            msg += create_user_list_html_string_for_message(new_user_profiles)
+            importer_log.add_success(msg)
+
+    return importer_log
+
+
+def normalize_rows(enrollment_rows: Iterable[EnrollmentParsedRow], importer_log: ImporterLog):
+    """The row schema has denormalized students and evaluations. Normalize / merge them back together"""
+    user_data_by_email: Dict[str, UserData] = {}
+    course_data_by_name_en: Dict[str, CourseData] = {}
+
+    for row in enrollment_rows:
+        stored = user_data_by_email.setdefault(row.student_data.email, row.student_data)
+        assert stored == row.student_data
+
+        stored = user_data_by_email.setdefault(row.responsible_data.email, row.responsible_data)
+        assert stored == row.responsible_data
+
+        course_data = course_data_by_name_en.setdefault(row.course_data.name_en, row.course_data)
+        assert course_data.equals_when_ignoring_degrees(row.course_data)
+
+        # Not a checker to keep merging and notifying about the merge together.
+        if not row.course_data.degrees.issubset(course_data.degrees):
+            course_data.degrees.update(row.course_data.degrees)
+
+            importer_log.add_warning(
+                _(
+                    "{location}: The course's \"{course_name}\" degree differs from it's degrees in previous rows."
+                    " All degrees have been set for the course."
```

Comment:
I would merge for this line on its own, on `main` it says "both" which is just factually incorrect

Reply:
(note to self: should be "its". Also the sentence structure is weird: The course's "introduction to X" degree differs [...].)